### PR TITLE
Implement KeyGenerator

### PR DESCRIPTION
**Pull request by contractormario**
PR adds `core/format/KeyGenerator`. Usage:

`const u64 = KeyGenerator.fromString('key_name') // returns UInt64`

### CODE REVIEW

File: src/core/format/KeyGenerator.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { UInt64 } from '../../model/UInt64';
+import { sha3_256 } from 'js-sha3';
+
+export class KeyGenerator {
+    /**
+     * @returns {UInt64} Deterministic uint64 value for the given string
```
Comment (rg911) [inline src/core/format/KeyGenerator.ts]: doc is wrong. missing description and and params

File: src/core/format/KeyGenerator.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+     */
+    public static fromString(input: string) {
+        if (input.length === 0) {
+            throw Error(`Input must not be empty`);
+        }
+        if (input.length > 1024) {
```
Comment (rg911) [inline src/core/format/KeyGenerator.ts]: not all keys have the same restriction. the check should be case specific

File: src/core/format/KeyGenerator.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            throw Error(`Input exceeds 1024 characters (has ${input.length})`);
+        }
+        const format = /^[a-zA-Z0-9_]+$/gm;
+        if (input.match(format) === null) {
+            throw Error(`Input has invalid format (accepted characters: a-z, A-Z, 0-9, _)`);
+        }
+        const hex = sha3_256(input)
```
Comment (rg911) [inline src/core/format/KeyGenerator.ts]: why hash the input string? should only need to buffer the string

File: src/core/format/KeyGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        return UInt64.fromHex(hex.substr(0, 16))
```
Comment (rg911) [inline src/core/format/KeyGenerator.ts]: should return hex string as per recent rest changes.

File: src/core/format/KeyGenerator.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+import { UInt64 } from '../../model/UInt64';
+import { sha3_256 } from 'js-sha3';
+
+export class KeyGenerator {
+    /**
+     * @returns {UInt64} Deterministic uint64 value for the given string
+     */
+    public static fromString(input: string) {
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply src/core/format/KeyGenerator.ts]: Missing return type in the signature

File: src/core/format/KeyGenerator.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        if (input.length === 0) {
+            throw Error(`Input must not be empty`);
+        }
+        if (input.length > 1024) {
+            throw Error(`Input exceeds 1024 characters (has ${input.length})`);
+        }
+        const format = /^[a-zA-Z0-9_]+$/gm;
```
[the remark this answers is not in the document]
Comment (contractormario) [reply src/core/format/KeyGenerator.ts]: Removed

File: src/core/format/KeyGenerator.ts (hunk 1 above)
Comment (contractormario) [reply src/core/format/KeyGenerator.ts]: Fixed

File: src/core/format/KeyGenerator.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2018 NEM
```
[the remark this answers is not in the document]
Comment (contractormario) [reply src/core/format/KeyGenerator.ts]: Fixed

File: src/core/format/KeyGenerator.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (contractormario) [reply src/core/format/KeyGenerator.ts]: What name do you suggest, generateUint64? generateKey?

File: src/core/format/KeyGenerator.ts (hunk 2 above)
Comment (contractormario) [reply src/core/format/KeyGenerator.ts]: Not sure what you mean?

File: src/core/format/KeyGenerator.ts (hunk 3 above)
Comment (contractormario) [reply src/core/format/KeyGenerator.ts]: longer strings with identical first 16 characters would produce the same u64. hashing guarantees uniqueness

File: src/core/format/KeyGenerator.ts (hunk 2 above)
Comment (rg911) [reply src/core/format/KeyGenerator.ts]: If I remember correctly only the `metadata.value` require this check. Don't think we need this check at all. do we?

File: src/core/format/KeyGenerator.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+     */
+    public static fromString(input: string): UInt64 {
+        if (input.length === 0) {
+            throw Error(`Input must not be empty`);
+        }
+        if (input.length > 1024) {
+            throw Error(`Input exceeds 1024 characters (has ${input.length})`);
+        }
+        const hex = sha3_256(input)
+        return UInt64.fromHex(hex.substr(0, 16))
```
Comment (rg911) [inline src/core/format/KeyGenerator.ts]: if you check the definition of UInt64, it uses 2 Uint32 (low / high). I would convert the hash buffer array to uint32 array and return the new Uint64( ['low', 'high']). Not quite sure about the substring here.

I would do something like
```
const result = new Uint32Array(hex.arrayBuffer());
return new UInt64([result[0], (result[1] | 0x80000000) >>> 0]);
```

File: src/core/format/KeyGenerator.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import { UInt64 } from '../../model/UInt64';
+import { sha3_256 } from 'js-sha3';
+
+export class KeyGenerator {
+    /**
+     * Generate UInt64 from a string
+     * @param {string} input Input string
+     * @returns {UInt64} Deterministic uint64 value for the given string
+     */
+    public static fromString(input: string): UInt64 {
```
Comment (rg911) [inline src/core/format/KeyGenerator.ts]: Shall we call it `generateUInt64Key` instead? 
@evias @dgarcia360

**Pull request merged**